internal/controller/supervisorconfig/oidcclientwatcher/oidc_client_watcher.go

```diff
@@ -0,0 +1,317 @@
+	allowedScopesValid     = "AllowedScopesValid"
+
+	reasonSuccess              = "Success"
+	reasonMissingRequiredValue = "MissingRequiredValue"
+	reasonNoClientSecretFound  = "NoClientSecretFound"
+
+	authorizationCodeGrantTypeName = "authorization_code"
+	refreshTokenGrantTypeName      = "refresh_token"
+	tokenExchangeGrantTypeName     = "urn:ietf:params:oauth:grant-type:token-exchange" //nolint:gosec // this is not a credential
+
+	openidScopeName          = "openid"
+	offlineAccessScopeName   = "offline_access"
+	requestAudienceScopeName = "pinniped:request-audience"
+	usernameScopeName        = "username"
+	groupsScopeName          = "groups"
+
+	allowedGrantTypesFieldName = "allowedGrantTypes"
+	allowedScopesFieldName     = "allowedScopes"
+
+	secretTypeToObserve = "storage.pinniped.dev/oidc-client-secret" //nolint:gosec // this is not a credential
+)
+
+type oidcClientWatcherController struct {
+	pinnipedClient     pinnipedclientset.Interface
+	oidcClientInformer configInformers.OIDCClientInformer
+	secretInformer     corev1informers.SecretInformer
+}
+
+// NewOIDCClientWatcherController returns a controllerlib.Controller that watches OIDCClients and updates
+// their status with validation errors.
+func NewOIDCClientWatcherController(
+	pinnipedClient pinnipedclientset.Interface,
+	secretInformer corev1informers.SecretInformer,
+	oidcClientInformer configInformers.OIDCClientInformer,
+	withInformer pinnipedcontroller.WithInformerOptionFunc,
+) controllerlib.Controller {
+	return controllerlib.New(
+		controllerlib.Config{
+			Name: "OIDCClientWatcherController",
+			Syncer: &oidcClientWatcherController{
+				pinnipedClient:     pinnipedClient,
+				secretInformer:     secretInformer,
+				oidcClientInformer: oidcClientInformer,
+			},
+		},
+		// We want to be notified when an OIDCClient's corresponding secret gets updated or deleted.
+		withInformer(
+			secretInformer,
+			pinnipedcontroller.MatchAnySecretOfTypeFilter(secretTypeToObserve, pinnipedcontroller.SingletonQueue()),
+			controllerlib.InformerOption{},
+		),
+		// We want to be notified when anything happens to an OIDCClient.
+		withInformer(
+			oidcClientInformer,
+			pinnipedcontroller.MatchAnythingFilter(pinnipedcontroller.SingletonQueue()),
+			controllerlib.InformerOption{},
+		),
+	)
+}
+
+// Sync implements controllerlib.Syncer.
+func (c *oidcClientWatcherController) Sync(ctx controllerlib.Context) error {
+	// Sync could be called on either a Secret or an OIDCClient, so to keep it simple, revalidate
+	// all OIDCClients whenever anything changes.
+	oidcClients, err := c.oidcClientInformer.Lister().List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to list OIDCClients: %w", err)
+	}
+
+	// We're only going to use storage to call GetName(), which happens to not need the constructor params.
+	// This is because we can read the Secrets from the informer cache here, instead of doing live reads.
+	storage := oidcclientsecretstorage.New(nil, nil)
```

Comment:
This is really weird constructor.  `GetName` should just be a free standing function.

Reply:
Certainly open to suggestions/refactors. This allowed some private things to remain private in the storage package and to keep the intimate knowledge of how the name is decided inside the the `crud` package.